### PR TITLE
Use a dedicated LDAP host and port for background jobs if configured

### PR DESCRIPTION
**Pull request by come-nc**
Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

### CODE REVIEW

File: apps/user_ldap/lib/Connection.php [hunk 1]
```diff
@@ -594,12 +594,18 @@ private function establishConnection() {
 
 			$isOverrideMainServer = ($this->configuration->ldapOverrideMainServer
 				|| $this->getFromCache('overrideMainServer'));
-			$isBackupHost = (trim($this->configuration->ldapBackupHost) !== "");
+			$isBackupHost = (trim($this->configuration->ldapBackupHost) !== "")
```
Comment (artonge) [inline apps/user_ldap/lib/Connection.php:597]: Would it make sense to rename to `hasBackupHost` to make it more obvious what this variable contains ?

File: apps/user_ldap/lib/Connection.php [hunk 2]
```diff
@@ -594,12 +594,18 @@ private function establishConnection() {
...
+				&& (!\OC::$CLI || !$this->configuration->ldapBackgroundHost);
```
Comment (artonge) [inline apps/user_ldap/lib/Connection.php:598]: Hard to read, and adds conditions unrelated to the backup host.
I would split into another variable and make the config check more explicit:

```suggestion
				$shouldUseBackgroundHost = \OC::$CLI && (trim($this->configuration->ldapBackgroundHost) !== "");
```

File: apps/user_ldap/lib/Connection.php [hunk 3]
```diff
@@ -611,7 +617,7 @@ private function establishConnection() {
 			//if LDAP server is not reachable, try the Backup (Replica!) Server
 			if ($isBackupHost || $isOverrideMainServer) {
```
Comment (artonge) [inline apps/user_ldap/lib/Connection.php:618]: And make this condition more explicit

```suggestion
			if (($isBackupHost || $isOverrideMainServer) && !shouldUseBackgroundHost) {
```

**Review by artonge — CHANGES_REQUESTED** [review]

**Comment by come-nc** [comment]
I pushed a commit to make code clearer, but it makes it clear that there is a logic flaw:

When in CLI, background host is used and backuphost is not used even if backgroundhost do not answer. But, if backuphost is used by web UI because a fallback, then it will be used in CLI as well because of the cache.

So we need to decide to either:

- [ ] Fallback to backuphost even for CLI if backgroundhost do not anwser
- [ ] Never use backuphost in CLI event if backgroundhost is not answering, we use no fallback

(These is considering there is a backgroundhost configured, if not we always fallback if needed to keep the same behavior as before)

**Comment by come-nc** [comment]
(There is also the question what to do when backup host is forced from admin settings, should that apply to cli if there is a background host?)

**Comment by blizzz** [comment]
>  Fallback to backuphost even for CLI if backgroundhost do not anwser

I would say so (provided that backuphost differs from background host)

**Comment by come-nc** [comment]
> > Fallback to backuphost even for CLI if backgroundhost do not anwser
> 
> I would say so (provided that backuphost differs from background host)

Ok, makes sense to me to always fallback to the backup if first host is not answering.
It also means forcing backup applies everywhere, which is more intuitive.

The problem with that is that we’ll need to split cache into 2 vars, to not go to backup for CLI when main is down but backgroundHost is still working.
I’ll implement that.

**Review by artonge — APPROVED** [review]

**Comment by come-nc** [comment]
rebased

**Review by blizzz — APPROVED** [review]
looks good

**Pull request merged**